src/storage.ts

```diff
@@ -0,0 +1,94 @@
+const store = (type: 'sessionStorage' | 'localStorage') => ({
+	available: undefined as boolean | undefined,
+	storage: window[type],
+	isAvailable(): boolean | undefined {
+		const key = 'local-storage-test';
+
+		if (this.available !== undefined) {
+			return this.available;
+		}
+
+		try {
+			// to fully test, need to set item
+			// http://stackoverflow.com/questions/9077101/iphone-localstorage-quota-exceeded-err-issue#answer-12976988
+			this.storage.setItem(key, 'graun');
+			this.storage.removeItem(key);
+			this.available = true;
+		} catch (err) {
+			this.available = false;
+		}
+
+		return this.available;
+	},
+
+	get(key: string): unknown | null {
+		if (!this.available) {
+			return;
+		}
+
+		let data;
+
+		// try and parse the data
+		try {
+			const value = this.getRaw(key);
+
+			if (value === null || value === undefined) {
+				return null;
+			}
+
+			data = JSON.parse(value);
+
+			if (data === null) {
+				return null;
+			}
+		} catch (e) {
+			this.remove(key);
```

Comment:
Is it slightly surprising behaviour that an operation named `get` could *delete* data?

Reply:
That. is. weird. Totally agree. I think I see the logic of it, it's deciding that the key is invalid and trying to clean up but there's a strong argument to say it's better to keep the invalid data to help with any debugging. I think it's reasonable to remove this step